cli/azd/pkg/azsdk/zip_deploy_client.go

```diff
@@ -113,8 +119,218 @@ func (c *ZipDeployClient) BeginDeploy(
 	return runtime.NewPoller(response, c.pipeline, pollerOptions)
 }
 
+// Deploys the specified application zip to the azure app service using deployment status api and waits for completion
+func (c *ZipDeployClient) BeginDeployTrackStatus(
+	ctx context.Context,
+	zipFile io.Reader,
+	subscriptionId,
+	resourceGroup,
+	appName string,
+) (*runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse], error) {
+	request, err := c.createDeployRequest(ctx, zipFile)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := c.pipeline.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	defer response.Body.Close()
+
+	if !runtime.HasStatusCode(response, http.StatusAccepted) {
+		return nil, runtime.NewResponseError(response)
+	}
+
+	client, err := armappservice.NewWebAppsClient(subscriptionId, c.cred, &arm.ClientOptions{
+		ClientOptions: policy.ClientOptions{
+			Retry: policy.RetryOptions{
+				MaxRetries: -1},
+		},
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("creating web app client: %w", err)
+	}
+
+	deploymentStatusId := response.Header.Get("Scm-Deployment-Id")
+	if deploymentStatusId == "" {
+		return nil, fmt.Errorf("empty deployment status id")
+	}
+
+	// nolint:lll
+	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
+	var poller *runtime.Poller[armappservice.WebAppsClientGetProductionSiteDeploymentStatusResponse]
+	err = retry.Do(
+		ctx,
+		retry.WithMaxRetries(3, retry.NewConstant(5*time.Second)),
+		func(ctx context.Context) error {
+			pollerRetry, err := client.BeginGetProductionSiteDeploymentStatus(ctx, resourceGroup, appName, deploymentStatusId, nil)
+			if err != nil {
+				var httpErr *azcore.ResponseError
+				if errors.As(err, &httpErr) {
+					// 404 - Retry if the resource is not found as deployment id doesn't match with temp deployment id
+					// Rest is the status code that runs default three retires in go sdk
+					if httpErr.StatusCode == 404 ||
+						httpErr.StatusCode == 408 ||
+						httpErr.StatusCode == 429 ||
+						httpErr.StatusCode == 500 ||
+						httpErr.StatusCode == 502 ||
+						httpErr.StatusCode == 503 ||
+						httpErr.StatusCode == 504 {
+						return retry.RetryableError(err)
+					}
+				}
+				return err
+			}
+
+			poller = pollerRetry
+
+			return nil
+		})
+
+	if err != nil {
+		return nil, fmt.Errorf("getting deployment status: %w", err)
+	}
```

Comment:
It may look a little cleaner if we use the retry options for `azure-sdk-for-go`. I believe the corresponding change would look like the following for your use case here. Note that we use `policy.WithRetryOptions` to create a `context.Context` that only affects the current request, and not the overall client.

```golang
	retryCtx := policy.WithRetryOptions(ctx, policy.RetryOptions{
		MaxRetries: 4,
		StatusCodes: append([]int{
			http.StatusRequestTimeout,      // 408
			http.StatusTooManyRequests,     // 429
			http.StatusInternalServerError, // 500
			http.StatusBadGateway,          // 502
			http.StatusServiceUnavailable,  // 503
			http.StatusGatewayTimeout,      // 504
		}, http.StatusNotFound), // 404
	})
	// nolint:lll
	// Example definition: https://github.com/Azure/azure-rest-api-specs/tree/main/specification/web/resource-manager/Microsoft.Web/stable/2022-03-01/examples/GetSiteDeploymentStatus.json
	poller, err := client.BeginGetProductionSiteDeploymentStatus(retryCtx, resourceGroup, appName, deploymentStatusId, nil)
	if err != nil {
		return nil, fmt.Errorf("getting deployment status: %w", err)
	}
```

Reply:
the call to `client.BeginGetProductionSiteDeploymentStatus` does not manual retries. It can use the default sdk client retries.  This API is not the one sending the zip file to Azure.

Reply:
We need manual retry for the call in line 101:
https://github.com/Azure/azure-dev/pull/3965/files/9ad7904c603b8e87a447789a02920b09651b6e2b#diff-33f0d8e97cd66b4e5eafa3c7ad76e09444aeedf30441eb8d993c18e4588be8e5R101

Reply:
FYI, while recently going through the implementation for `api/publish`, I did discover that `azure-sdk-for-go` handles retries for streaming body content perfectly fine. We just need to pass an `io.ReadSeekCloser` and use `request.SetBody`. You can see it working [here](https://github.com/Azure/azure-dev/pull/3923/files#diff-33f0d8e97cd66b4e5eafa3c7ad76e09444aeedf30441eb8d993c18e4588be8e5). After this lands, I'll fix the zip deploy logic here.